receiver/googlecloudspannerreceiver/internal/filter/itemcardinality.go

```diff
@@ -0,0 +1,183 @@
+	totalLimit         int
+	limitByTimestamp   int
+	itemActivityPeriod time.Duration
+	logger             *zap.Logger
+	cache              *ttlcache.Cache
+}
+
+type currentLimitByTimestamp struct {
+	limitByTimestamp int
+}
+
+func (f *currentLimitByTimestamp) dec() {
+	f.limitByTimestamp--
+}
+
+func (f *currentLimitByTimestamp) get() int {
+	return f.limitByTimestamp
+}
+
+func NewItemCardinalityFilter(metricName string, totalLimit int, limitByTimestamp int,
+	itemActivityPeriod time.Duration, logger *zap.Logger) (ItemFilter, error) {
+	if limitByTimestamp > totalLimit {
+		return nil, fmt.Errorf("total limit %q is lower or equal to limit by timestamp %q", totalLimit, limitByTimestamp)
+	}
+
+	cache := ttlcache.NewCache()
+
+	cache.SetCacheSizeLimit(totalLimit)
+	cache.SkipTTLExtensionOnHit(true)
+
+	return &itemCardinalityFilter{
+		metricName:         metricName,
+		totalLimit:         totalLimit,
+		limitByTimestamp:   limitByTimestamp,
+		itemActivityPeriod: itemActivityPeriod,
+		logger:             logger,
+		cache:              cache,
+	}, nil
+}
+
+func (f *itemCardinalityFilter) TotalLimit() int {
+	return f.totalLimit
+}
+
+func (f *itemCardinalityFilter) LimitByTimestamp() int {
+	return f.limitByTimestamp
+}
+
+func (f *itemCardinalityFilter) Filter(sourceItems []*Item) ([]*Item, error) {
+	var filteredItems []*Item
+	groupedItems := groupByTimestamp(sourceItems)
+	sortedItemKeys := sortedKeys(groupedItems)
+
+	for _, key := range sortedItemKeys {
+		filteredGroupedItems, err := f.filterItems(groupedItems[key])
+		if err != nil {
+			return nil, err
+		}
+
+		filteredItems = append(filteredItems, filteredGroupedItems...)
+	}
+
+	return filteredItems, nil
+}
+
+func (f *itemCardinalityFilter) filterItems(items []*Item) ([]*Item, error) {
+	limit := currentLimitByTimestamp{
+		limitByTimestamp: f.limitByTimestamp,
+	}
+
+	var filteredItems []*Item
+	for _, item := range items {
+		if included, err := f.includeItem(item, &limit); err != nil {
+			return nil, err
+		} else if included {
+			filteredItems = append(filteredItems, item)
+		}
+	}
+
+	return filteredItems, nil
+}
+
+func (f *itemCardinalityFilter) includeItem(item *Item, limit *currentLimitByTimestamp) (bool, error) {
+	if _, err := f.cache.Get(item.SeriesKey); err == nil {
+		return true, nil
+	} else if err != ttlcache.ErrNotFound {
+		return false, err
+	}
+
+	if !f.canIncludeNewItem(limit.get()) {
+		f.logger.Debug("Skip item", zap.String("seriesKey", item.SeriesKey), zap.Time("timestamp", item.Timestamp))
+		return false, nil
+	}
+
+	if err := f.cache.SetWithTTL(item.SeriesKey, struct{}{}, f.itemActivityPeriod); err != nil {
+		if err == ttlcache.ErrClosed {
+			err = fmt.Errorf("set item from cache failed for metric %q because cache has been already closed: %w", f.metricName, err)
+		}
+		return false, err
+	}
+
+	f.logger.Debug("Added item to cache", zap.String("seriesKey", item.SeriesKey), zap.Time("timestamp", item.Timestamp))
+
+	limit.dec()
```

Comment:
Are you sure this isn't called concurrently? If you are receiving concurrent metrics, I would assume this is then called concurrently as well.

Reply:
You can inspect receiver for reading metrics method. It calls appropriate Read() method of each reader sequentially, then metric data points are collected together and grouped by metric - no concurrency there. Then cardinality flitering will be applied to each of metric sequentially (one instance of filter per each metric) - you will see this in second MR.